src/sa.py

```diff
@@ -7,8 +14,187 @@ def main():
+     -alphabet size is bounded by n anyway.
+    Buckets constructed as dict, in order of alphabet string. O(n)
+    Input string is split into suffixes, adding $ so each has length n, O(n*n)
+     -could get rid of slicing and just use index, but am lazy, and already implemented radix using strings (and we're n^2 anyway)
+    radix sort suffixes O(n*n)
+    one last O(n) running through sorted suffixes to get their starting index (because lazyness.)
+    '''
+    n = len(x)
+    alpha = alphabet_from_input_string(x)
+    #suffixes = [x[i:]+i*'$' for i, _ in enumerate(x)]
+    x += '$'
+    suffixes = [suffix(x, i) for i, _ in enumerate(x)]
+
+    sort = radix_sort(suffixes, alpha)
+    return [x.i for x in sort]
+
+
+def sa_construction_simple_nsq(x: str) -> list[int]:
+    '''
+    simple sa construction uing builtin strcmp and sort, for comparisson
+    '''
+    suf = [x[i:] for i, _ in enumerate(x)]
+    suf = sorted(suf)
+    return [len(x) - len(i) for i in suf]
+
+
+def lower_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for lower bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if x[(sa[mid] + i) % len(x)] < a:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def upper_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for upper bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if ord(x[(sa[mid] + i) % len(x)]) < ord(a)+1:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def pattern_match(x: str, p: str, sa: list[int]) -> Iterable[int]:
+    ''' report all occurences of pattern p in string x, given Suffix array: SA of x. '''
+    high = len(sa)
+    low = 0
+
+    if len(p) == 0 or len(x) == 0:
+        return []
+    for i, a in enumerate(p):
+        low = lower_bound(a, i, low, high, x, sa)
+        high = upper_bound(a, i, low, high, x, sa)
+
+    return sa[low:high]
+
+
+# Work in progress below
+def radix_by_index(x: str, u: list[int], alphabet: list[str]) -> dict[str, str]:
+    buckets = {i: 0 for i in alphabet}
+    triplets = get_triplet_map(x, u)
+
+    arr = [i for i in u]  # make copy so we don't override u
+    sa = [0] * len(u)
+    for i in range(3):
+        for trip in arr:
+            buckets[triplets[trip][i]] += 1
+        accsum = 0
+        for bucket in buckets:
+            buckets[bucket], accsum = accsum, accsum + buckets[bucket]
+        for trip in arr:
+            sa[buckets[triplets[trip][i]]] = trip
+            buckets[triplets[trip][i]] += 1
+        for bucket in buckets:
+            buckets[bucket] = 0
+        arr, sa = sa, arr
```

Comment:
I assume that the triplets are reversed? You need to sort least-significant-first to get the sorted, after all.